File: neetcode-150-course/01-arrays-hashing/code/LabValidAnagram.java
// Lab 01E — Valid Anagram (reference solution).
//
// Run:
//     mkdir -p /tmp/out
//     javac -d /tmp/out 01-arrays-hashing/code/LabValidAnagram.java
//     java -ea -cp /tmp/out LabValidAnagram
public class LabValidAnagram {
    public static boolean isAnagram(String s, String t) {
        if (s.length() != t.length()) return false;
        int[] count = new int[26];
        for (int i = 0; i < s.length(); i++) {
            count[s.charAt(i) - 'a']++;
            count[t.charAt(i) - 'a']--;
        }
        for (int c : count) if (c != 0) return false;
        return true;
    }

    public static void main(String[] args) {
        assert isAnagram("anagram", "nagaram");
        assert !isAnagram("rat", "car");
        assert isAnagram("a", "a");
        assert !isAnagram("ab", "a");
        assert isAnagram("", "");
        System.out.println("all tests passed");
    }
}

File: neetcode-150-course/01-arrays-hashing/code/lab_valid_anagram.py
"""Lab 01D — Valid Anagram (reference solution)."""


def is_anagram(s: str, t: str) -> bool:
    from collections import Counter
    return Counter(s) == Counter(t)


if __name__ == "__main__":
    assert is_anagram("anagram", "nagaram") is True
    assert is_anagram("rat", "car") is False
    assert is_anagram("a", "a") is True
    assert is_anagram("ab", "a") is False
    assert is_anagram("", "") is True
    print("all tests passed")
